hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java

```diff
@@ -139,6 +140,8 @@ protected void doRefresh() {
 
   @Override
   protected void doCommit(TableMetadata base, TableMetadata metadata) {
+    boolean updateTable = base != null || metadata.propertyAsBoolean(TABLE_FROM_HIVE, false);
```

Comment:
Is this telling Iceberg to overwrite the Hive table? I think this property should probably not be persisted to Hive in table properties.

Reply:
If we do not want to use properties for signaling this behavior (looking at it again with fresh eyes seems like a hack), then we have to modify:
- add method TableBuilder.updateExisting
- add method TableOperations.commit with new parameter signaling update
- add method HiveTableOperations/HadoopTableOperations to signal the update

Or we have to resort to parameters sending/hacking through some level and removing the parameter from the lower levels.

I do not like any of these solutions, so happy to go with any other viable solution...

Reply:
What is the purpose of this? Is it because the table already exists in the Hive MetaStore but there is no Iceberg metadata?

Reply:
Yes, you are right.

With the HiveMetaHooks we have the possibility to do some `preCreate` and `commitCreate` stuff, but the table should be created by the HMS between these phases. I have even thought about dropping the HMS created table in the `commitCreate` method but that seems like a serious waste of effort, and also I do not have a HMS client at hand.

Reply:
Okay, thanks for the context.

I think we may be able to catch this case within `HiveTableOperations`. When creating a table, we will create a new `TableOperations` and that will attempt to load the current metadata. Normally, [that will fail](https://github.com/apache/iceberg/blob/master/hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java#L125) and because the `currentMetadataLocation` was `null`, the failure is caught and metadata is set to null because the table doesn't exist. If the table exists but doesn't have a metadata location, then we will get a table from Hive, but the location property won't exist. We can detect that the location property doesn't exist and set a flag, `updateToCreate`.

That avoids the need for a table property because we detect the state of the Hive MetaStore, and it doesn't incur extra thrift calls.

Reply:
This would be a slight modification to the API.
- Before the change we throw a NoSuchIcebergTableException if we tried to load a Hive table which did not have the correct table type and the metadata location set.
- After the change we will try to update the table with the correct parameters

Would this be acceptable change in the API behavior?

Reply:
Could we avoid the behavior change by adding a property in the pre-create hook? If we signal that Hive is creating the table, then this could set the flag to update internally, do the update, and drop the flag from Hive metadata. It still avoids additional calls to Hive.

Reply:
This flag is the HiveTableOperations#TABLE_FROM_HIVE ("iceberg.table.from.hive").
We do not store this to the HMS, but we store this to the Iceberg table since in the HiveTableOperations.doCommit(base, metadata) the metadata is Immutable.

Reply:
I agree that this should not change the `TableMetadata` that was passed in. If we pass the flag in `TableMetadata`, then it would be stored in the Iceberg metadata.

But why can't we add this flag to the HMS table that is passed to the pre-commit phase in the Hive hooks? If we add it to the Hive table that gets created and should be replaced, then there is no need to pass it in Iceberg metadata. And we can remove it when we create the Iceberg table underneath and set the metadata location.

Reply:
Oh... now I get it.
Seems like a good idea, I will check it!

Reply:
Actually, I see that we already set the table type property in the pre-commit hook. We could use that to detect this case and avoid throwing `NoSuchIcebergTableException`, then use update.

Reply:
Uploaded the change.
We need to call HMSClient.getTable every time in HiveTableOperations.doCommit, so we can decide if this is a new table or already created by the Hook.
This is an extra call in case the new table creation is started from Spark. This might be acceptable.